pytket/phir/phirgen_parallel.py

```diff
@@ -0,0 +1,317 @@
+##############################################################################
+#
+# Copyright (c) 2023 Quantinuum LLC All rights reserved.
+# Use of this source code is governed by a BSD-style
+# license that can be found in the LICENSE file.
+#
+##############################################################################
+
+# mypy: disable-error-code="misc"
+
+import json
+import logging
+from typing import TYPE_CHECKING, Any
+
+import pytket.circuit as tk
+from phir.model import PHIRModel
+from pytket.unit_id import UnitID
+
+from .machine import Machine
+from .phirgen import append_cmd, arg_to_bit, get_decls, tket_gate_to_phir
+from .sharding.shard import Cost, Ordering, Shard, ShardLayer
+
+if TYPE_CHECKING:
+    import sympy
+
+logger = logging.getLogger(__name__)
+
+
+def process_sub_commands(
+    sub_commands: dict[UnitID, list[tk.Command]], max_parallel_sq_gates: int
+) -> dict[int, list[tk.Command]]:
+    """Create parallelizable groups of sub-commands."""
+    groups: dict[
+        int, list[tk.Command]
+    ] = {}  # be sure to order by group number into a list when returning
+    qubits2groups = {}  # track the most recent group in which a qubit was used
+    # group numbers for each gate are incremented by 2 so they don't overlap
+    # and different gate types don't go in the same group
+    # RZ gates go in %3=1 groups, R1XY gates go in %3=1 groups,
+    # and all other gates will go in %3=2 groups
+    rz_group_number = -3  # will be set to 0 when first RZ gate is assigned (-3 + 3 = 0)
+    r1xy_group_number = (
+        -2  # will be set to 1 when first R1XY gate is assigned (-2 + 3 = 1)
+    )
+    other_group_number = (
+        -1  # will be set to 2 when first R1XY gate is assigned (-1 + 3 = 2)
+    )
+    num_scs_per_qubit = {}
+    index = 0
+
+    for qubit in sub_commands:
+        num_scs_per_qubit[qubit] = len(sub_commands[qubit])
+        # set every qubits value to be -4
+        # prevents KeyError in check for group number
+        # will get set to a valid group number the first time the qubit is used
+        qubits2groups[qubit] = -4
+    max_len = max(num_scs_per_qubit.values())
+
+    while index < max_len:
```

Comment:
Does it make sense to write this loop as follows?
```suggestion
    for index in range(max_len):
```
and delete lines 49 and 104?